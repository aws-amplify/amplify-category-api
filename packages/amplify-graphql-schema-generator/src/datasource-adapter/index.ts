export * from './datasource-adapter';
export * from './string-datasource-adapter';
export { MySQLDataSourceAdapter } from './mysql-datasource-adapter';
export { PostgresDataSourceAdapter } from './pg-datasource-adapter';
export { MySQLStringDataSourceAdapter } from './mysql-string-datasource-adapter';
export { PostgresStringDataSourceAdapter } from './pg-string-datasource-adapter';
