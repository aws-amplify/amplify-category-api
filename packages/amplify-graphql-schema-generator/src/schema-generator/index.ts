export * from './generate-schema';
export * from './schema-overrides';
