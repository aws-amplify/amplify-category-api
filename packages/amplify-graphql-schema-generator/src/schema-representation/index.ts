export * from './engine';
export * from './schema';
export * from './types';
