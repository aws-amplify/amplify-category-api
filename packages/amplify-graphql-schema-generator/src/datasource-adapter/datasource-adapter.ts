import * as os from 'os';
import { Field, Index, Model } from '../schema-representation';

export interface DataSourceConfig {
  host: string;
  port: number;
  database: string;
  username: string;
  password: string;
  sslCertificate?: string;
}

export abstract class DataSourceAdapter {
  public abstract getTablesList(): string[];

  public abstract getFields(tableName: string): Field[];

  public abstract getPrimaryKey(tableName: string): Index | null;

  public abstract getIndexes(tableName: string): Index[];

  public abstract initialize(): Promise<void>;

  public abstract cleanup(): void;

  public abstract test(): Promise<boolean>;

  protected abstract querySchema(): Promise<string>;

  public useVPC = false;

  public vpcSchemaInspectorLambda: string | undefined = undefined;

  public vpcLambdaRegion: string | undefined = undefined;

  public getModels(): Model[] {
    const tableNames = this.getTablesList();
    const models = [];
    for (const table of tableNames) {
      models.push(this.describeTable(table));
    }
    return models;
  }

  public describeTable(tableName: string): Model {
    // Retrieve the fields, primary key and indexes info
    const fields = this.getFields(tableName);
    const primaryKey = this.getPrimaryKey(tableName);
    const indexes = this.getIndexes(tableName);

    // Construct the model from the retrieved details
    const model = new Model(tableName);
    fields.forEach((field) => model.addField(field));
    primaryKey && model.setPrimaryKey(primaryKey.getFields());
    indexes.forEach((index) => model.addIndex(index.name, index.getFields()));
    return model;
  }

  public useVpc(vpcSchemaInspectorLambda: string, region: string): void {
    this.useVPC = true;
    this.vpcSchemaInspectorLambda = vpcSchemaInspectorLambda;
    this.vpcLambdaRegion = region;
  }

  protected queryToCSV(queryResult: any[]): string {
    if (queryResult.length === 0) {
      return '';
    }
    const headers = Object.keys(queryResult[0]);
    const headerIndices = Object.fromEntries(headers.map((key, index) => [index, key]));
    const rows = queryResult.map((row) =>
      [...Array(headers.length).keys()]
        .map((index) => {
          const value = row[headerIndices[index]];
          // sanitize if comma is present in value
          if (typeof value === 'string' && value.includes(',')) {
            return `"${value}"`;
          }
          if (value === null) {
            return 'NULL';
          }
          return value;
        })
        .join(','),
    );
    return headers.join(',') + os.EOL + rows.join(os.EOL);
  }
}
