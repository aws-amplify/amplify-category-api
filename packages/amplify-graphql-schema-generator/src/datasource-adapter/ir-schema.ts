import { Field, Index } from '../schema-representation';

export type IRSchemaInputs = {
  fields: any;
  indexes: any;
  tables: any;
};

export abstract class IRSchema {
  protected abstract setFields(fields: string): void;
  protected abstract setIndexes(indexes: string): void;
  protected abstract setTables(tables: string): void;
  public abstract getFields(tableName: string): Field[];
  public abstract getPrimaryKey(tableName: string): Index | null;
  public abstract getIndexes(tableName: string): Index[];

  constructor({ fields, indexes, tables }: IRSchemaInputs) {
    this.setFields(fields);
    this.setIndexes(indexes);
    this.setTables(tables);
  }
}
