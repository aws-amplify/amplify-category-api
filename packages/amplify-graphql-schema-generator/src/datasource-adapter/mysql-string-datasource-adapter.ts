import { EnumType, Field, FieldDataType, FieldType, Index } from '../schema-representation';
import { StringDataSourceAdapter, EmptySchemaError, InvalidSchemaError } from './string-datasource-adapter';

export interface MySQLIndex {
  tableName: string;
  nonUnique: number;
  indexName: string;
  sequence: number;
  columnName: string;
  nullable: boolean;
}

export interface MySQLColumn {
  tableName: string;
  columnName: string;
  sequence: number;
  default: string;
  datatype: string;
  columnType: string;
  nullable: boolean;
  length: number | null | undefined;
}

type MySQLSchema = MySQLSchemaField[];

type MySQLSchemaField = {
  TABLE_NAME: string;
  COLUMN_NAME: string;
  COLUMN_DEFAULT: string;
  ORDINAL_POSITION: string;
  DATA_TYPE: string;
  COLUMN_TYPE: string;
  IS_NULLABLE: string;
  CHARACTER_MAXIMUM_LENGTH: string;
  INDEX_NAME: string;
  NON_UNIQUE: string;
  SEQ_IN_INDEX: string;
  NULLABLE: string;
};
const expectedColumns = [
  'TABLE_NAME',
  'COLUMN_NAME',
  'COLUMN_DEFAULT',
  'ORDINAL_POSITION',
  'DATA_TYPE',
  'COLUMN_TYPE',
  'IS_NULLABLE',
  'CHARACTER_MAXIMUM_LENGTH',
  'INDEX_NAME',
  'NON_UNIQUE',
  'SEQ_IN_INDEX',
  'NULLABLE',
];

export class MySQLStringDataSourceAdapter extends StringDataSourceAdapter {
  private dbBuilder: any;

  private indexes: MySQLIndex[];

  private fields: MySQLColumn[];

  private tables: string[];

  private enums: Map<string, EnumType> = new Map<string, EnumType>();

  private readonly PRIMARY_KEY_INDEX_NAME = 'PRIMARY';

  protected setSchema(schema: MySQLSchema): void {
    this.setFields(schema);
    this.setIndexes(schema);
    this.setTables(schema);
  }

  protected validateSchema(schema: any[]): schema is MySQLSchema {
    if (schema.length === 0) {
      throw new EmptySchemaError();
    }
    const columns = Object.keys(schema[0]);
    const hasAllExpectedColumns = expectedColumns.every((column) => columns.includes(column));
    if (!hasAllExpectedColumns) {
      throw new InvalidSchemaError(schema, expectedColumns);
    }

    return true;
  }

  protected setFields(parsedSchema: MySQLSchema): void {
    this.fields = parsedSchema.map((item: any) => ({
      tableName: item.TABLE_NAME,
      columnName: item.COLUMN_NAME,
      default: item.COLUMN_DEFAULT,
      sequence: item.ORDINAL_POSITION,
      datatype: item.DATA_TYPE,
      columnType: item.COLUMN_TYPE,
      nullable: item.IS_NULLABLE === 'YES',
      length: item.CHARACTER_MAXIMUM_LENGTH,
    }));
  }

  protected setIndexes(parsedSchema: MySQLSchema): void {
    this.indexes = parsedSchema
      .filter(({ INDEX_NAME }) => !!INDEX_NAME)
      .map((item: any) => ({
        tableName: item.TABLE_NAME,
        indexName: item.INDEX_NAME,
        nonUnique: item.NON_UNIQUE,
        columnName: item.COLUMN_NAME,
        sequence: item.SEQ_IN_INDEX,
        nullable: item.NULLABLE === 'YES',
      }));
  }

  protected setTables(parsedSchema: MySQLSchema): void {
    this.tables = Array.from(new Set(parsedSchema.map(({ TABLE_NAME }) => TABLE_NAME)));
  }

  public getTablesList(): string[] {
    return this.tables;
  }

  public getFields(tableName: string): Field[] {
    const fieldsName: string[] = [
      ...new Set(
        this.fields
          .filter((f) => f.tableName === tableName)
          .sort((a, b) => a.sequence - b.sequence)
          .map((f) => f.columnName),
      ),
    ];

    const modelFields = fieldsName.map((columnName) => {
      const dbField = this.fields.find((field) => field.tableName === tableName && field.columnName === columnName)!;
      const field: Field = {
        name: dbField.columnName,
        type: this.mapDataType(dbField.datatype, dbField.nullable, tableName, dbField.columnName, dbField.columnType),
        length: dbField.length,
        default: dbField.default
          ? {
              kind: 'DB_GENERATED',
              value: dbField.default,
            }
          : undefined,
      };
      return field;
    });

    return modelFields;
  }

  public getPrimaryKey(tableName: string): Index | null {
    const key = this.indexes
      .filter((index) => index.tableName === tableName && index.indexName === this.PRIMARY_KEY_INDEX_NAME)
      .sort((a, b) => a.sequence - b.sequence);

    if (!key || key.length == 0) {
      return null;
    }

    const index: Index = new Index(key[0].indexName);
    index.setFields(key.map((k) => k.columnName));

    return index;
  }

  public getIndexes(tableName: string): Index[] {
    const indexNames: string[] = [
      ...new Set(
        this.indexes.filter((i) => i.tableName === tableName && i.indexName !== this.PRIMARY_KEY_INDEX_NAME).map((i) => i.indexName),
      ),
    ];

    const tableIndexes = indexNames.map((indexName: string) => {
      const key = this.indexes
        .filter((index) => index.tableName == tableName && index.indexName === indexName)
        .sort((a, b) => a.sequence - b.sequence);
      const index: Index = new Index(indexName);
      index.setFields(key.map((k) => k.columnName));
      return index;
    });

    return tableIndexes;
  }

  public mapDataType(datatype: string, nullable: boolean, tableName: string, fieldName: string, columntype: string): FieldType {
    let fieldDatatype: FieldDataType = 'String';
    let listtype = false;

    switch (datatype.toUpperCase()) {
      case 'VARCHAR':
      case 'CHAR':
      case 'BINARY':
      case 'VARBINARY':
      case 'TINYBLOB':
      case 'TINYTEXT':
      case 'TEXT':
      case 'BLOB':
      case 'MEDIUMTEXT':
      case 'MEDIUMBLOB':
      case 'LONGTEXT':
      case 'LONGBLOB':
        fieldDatatype = 'String';
        break;
      case 'SET':
        fieldDatatype = 'String';
        listtype = true;
        break;
      case 'BOOLEAN':
      case 'BOOL':
        fieldDatatype = 'Boolean';
        break;
      case 'BIT':
      case 'TINYINT':
      case 'SMALLINT':
      case 'MEDIUMINT':
      case 'INT':
      case 'INTEGER':
      case 'BIGINT':
      case 'YEAR':
        fieldDatatype = 'Int';
        break;
      case 'FLOAT':
      case 'DOUBLE':
      case 'DECIMAL':
      case 'DEC':
      case 'NUMERIC':
        fieldDatatype = 'Float';
        break;
      case 'DATE':
        fieldDatatype = 'AWSDate';
        break;
      case 'TIMESTAMP':
      case 'DATETIME':
        fieldDatatype = 'AWSDateTime';
        break;
      case 'TIME':
        fieldDatatype = 'AWSTime';
        break;
      case 'JSON':
        fieldDatatype = 'AWSJSON';
        break;
      case 'ENUM':
        fieldDatatype = 'ENUM';
        break;
      default:
        fieldDatatype = 'String';
        break;
    }

    let result: FieldType;
    if (fieldDatatype === 'ENUM') {
      const enumName = this.generateEnumName(tableName, fieldName);
      result = {
        kind: 'Enum',
        values: this.getEnumValues(columntype),
        name: this.generateEnumName(tableName, fieldName),
      };
      this.enums.set(enumName, result);
    } else {
      result = {
        kind: 'Scalar',
        name: fieldDatatype,
      };
    }

    if (!nullable) {
      result = {
        kind: 'NonNull',
        type: result,
      };
    }

    return result;
  }

  private getEnumValues(value: string): string[] {
    // RegEx matches strings with quotes 'match' or "match"
    const regex = /(["'])(?:(?=(\\?))\2.)*?\1/g;
    // Remove the first and last character from the matched string which contains the quote
    return value.match(regex).map((a) => a.slice(1, -1));
  }

  private generateEnumName(tableName: string, fieldName: string) {
    const enumNamePrefix = [tableName, fieldName].join('_');
    let enumName = enumNamePrefix;
    let counter = 0;
    while (this.enums.has(enumName)) {
      enumName = [enumNamePrefix, counter.toString()].join('_');
      counter++;
    }
    return enumName;
  }
}
