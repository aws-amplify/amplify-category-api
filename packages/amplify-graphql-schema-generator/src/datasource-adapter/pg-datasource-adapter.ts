import { knex } from 'knex';
import { invokeSchemaInspectorLambda } from '../utils/vpc-helper';
import ora from 'ora';
import { EnumType, Field, FieldDataType, FieldType, Index } from '../schema-representation';
import { DataSourceAdapter } from './datasource-adapter';
import { PostgresStringDataSourceAdapter, expectedColumns } from './pg-string-datasource-adapter';

const spinner = ora();

export interface PostgresDataSourceConfig {
  host: string;
  port: number;
  database: string;
  username: string;
  password: string;
}

export class PostgresDataSourceAdapter extends DataSourceAdapter {
  private adapter: PostgresStringDataSourceAdapter;

  private dbBuilder: any;

  constructor(private config: PostgresDataSourceConfig) {
    super();
  }

  public async test(): Promise<boolean> {
    const TEST_QUERY = 'SELECT 1';
    if (!this.dbBuilder) {
      this.establishDBConnection();
    }
    try {
      await this.dbBuilder.raw(TEST_QUERY);
    } catch (error) {
      return false;
    }
    return true;
  }

  public async initialize(): Promise<void> {
    spinner.start('Fetching the database schema...');
    try {
      await this.establishDBConnection();
      const schema = await this.querySchema();
      this.adapter = new PostgresStringDataSourceAdapter(schema);
    } catch (error) {
      spinner.fail('Failed to fetch the database schema.');
      throw error;
    }
    spinner.succeed('Successfully fetched the database schema.');
  }

  private establishDBConnection(): void {
    const databaseConfig = {
      host: this.config.host,
      database: this.config.database,
      port: this.config.port,
      user: this.config.username,
      password: this.config.password,
      ssl: { rejectUnauthorized: false },
    };
    try {
      this.dbBuilder = knex({
        client: 'pg',
        connection: databaseConfig,
        pool: {
          min: 1,
          max: 1,
          createTimeoutMillis: 30000,
          acquireTimeoutMillis: 30000,
          idleTimeoutMillis: 30000,
          reapIntervalMillis: 1000,
          createRetryIntervalMillis: 100,
        },
        debug: false,
      });
    } catch (err) {
      console.info(err);
      throw err;
    }
  }

  public getTablesList(): string[] {
    return this.adapter.getTablesList();
  }

  protected async querySchema(): Promise<string> {
    const schemaQuery = getPostgresSchemaQuery(this.config.database);
    const result =
      this.useVPC && this.vpcSchemaInspectorLambda
        ? await invokeSchemaInspectorLambda(this.vpcSchemaInspectorLambda, this.config, schemaQuery, this.vpcLambdaRegion)
        : (await this.dbBuilder.raw(schemaQuery)).rows;

    return this.queryToCSV(result);
  }

  public getFields(tableName: string): Field[] {
    return this.adapter.getFields(tableName);
  }

  public getPrimaryKey(tableName: string): Index | null {
    return this.adapter.getPrimaryKey(tableName);
  }

  public getIndexes(tableName: string): Index[] {
    return this.adapter.getIndexes(tableName);
  }

  public cleanup(): void {
    this.dbBuilder && this.dbBuilder.destroy();
  }
}

export function getPostgresSchemaQuery(databaseName: string): string {
  return `
SELECT DISTINCT
  INFORMATION_SCHEMA.COLUMNS.table_name,
  ${expectedColumns.filter((column) => !(column === 'index_columns' || column === 'table_name')).join(',')},
  REPLACE(SUBSTRING(indexdef from '\\((.*)\\)'), '"', '') as index_columns
FROM INFORMATION_SCHEMA.COLUMNS
LEFT JOIN pg_indexes
ON
  INFORMATION_SCHEMA.COLUMNS.table_name = pg_indexes.tablename
  AND INFORMATION_SCHEMA.COLUMNS.column_name = ANY(STRING_TO_ARRAY(REPLACE(SUBSTRING(indexdef from '\\((.*)\\)'), '"', ''), ', '))
  LEFT JOIN (
    SELECT
      t.typname AS enum_name,
      ARRAY_AGG(e.enumlabel) as enum_values
    FROM    pg_type t JOIN
      pg_enum e ON t.oid = e.enumtypid JOIN
      pg_catalog.pg_namespace n ON n.oid = t.typnamespace
    WHERE   n.nspname = 'public'
    GROUP BY enum_name
  ) enums
  ON enums.enum_name = INFORMATION_SCHEMA.COLUMNS.udt_name
  LEFT JOIN information_schema.table_constraints
  ON INFORMATION_SCHEMA.table_constraints.constraint_name = indexname
  AND INFORMATION_SCHEMA.COLUMNS.table_name = INFORMATION_SCHEMA.table_constraints.table_name
WHERE INFORMATION_SCHEMA.COLUMNS.table_schema = 'public' AND INFORMATION_SCHEMA.COLUMNS.TABLE_CATALOG = '${databaseName}';
`;
}
