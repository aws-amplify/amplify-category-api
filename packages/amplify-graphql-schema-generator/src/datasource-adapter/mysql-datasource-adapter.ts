import { knex } from 'knex';
import ora from 'ora';
import { invokeSchemaInspectorLambda } from '../utils/vpc-helper';
import { Field, Index } from '../schema-representation';
import { getSSLConfig } from '../utils';
import { DataSourceAdapter } from './datasource-adapter';
import { MySQLStringDataSourceAdapter } from './mysql-string-datasource-adapter';

const spinner = ora();
export interface MySQLDataSourceConfig {
  host: string;
  port: number;
  database: string;
  username: string;
  password: string;
}

export class MySQLDataSourceAdapter extends DataSourceAdapter {
  private adapter: MySQLStringDataSourceAdapter;
  private dbBuilder: any;

  constructor(private config: MySQLDataSourceConfig) {
    super();
  }

  public async test(): Promise<boolean> {
    const TEST_QUERY = 'SELECT 1';
    if (!this.dbBuilder) {
      this.establishDBConnection();
    }
    try {
      await this.dbBuilder.raw(TEST_QUERY);
    } catch (error) {
      return false;
    }
    return true;
  }

  public async initialize(): Promise<void> {
    spinner.start('Fetching the database schema...');
    try {
      await this.establishDBConnection();
      const schema = await this.querySchema();
      this.adapter = new MySQLStringDataSourceAdapter(schema);
    } catch (error) {
      spinner.fail('Failed to fetch the database schema.');
      throw error;
    }
    spinner.succeed('Successfully fetched the database schema.');
  }

  private establishDBConnection(): void {
    const databaseConfig = {
      host: this.config.host,
      database: this.config.database,
      port: this.config.port,
      user: this.config.username,
      password: this.config.password,
      ssl: getSSLConfig(this.config.host),
    };
    try {
      this.dbBuilder = knex({
        client: 'mysql2',
        connection: databaseConfig,
        pool: {
          min: 5,
          max: 30,
          createTimeoutMillis: 30000,
          acquireTimeoutMillis: 30000,
          idleTimeoutMillis: 30000,
          reapIntervalMillis: 1000,
          createRetryIntervalMillis: 100,
        },
        debug: false,
      });
    } catch (err) {
      console.info(err);
      throw err;
    }
  }

  public getTablesList(): string[] {
    return this.adapter.getTablesList();
  }

  public getFields(tableName: string): Field[] {
    return this.adapter.getFields(tableName);
  }

  public getPrimaryKey(tableName: string): Index | null {
    return this.adapter.getPrimaryKey(tableName);
  }

  public getIndexes(tableName: string): Index[] {
    return this.adapter.getIndexes(tableName);
  }

  protected async querySchema(): Promise<string> {
    const schemaQuery = getMySQLSchemaQuery(this.config.database);
    const result =
      this.useVPC && this.vpcSchemaInspectorLambda
        ? await invokeSchemaInspectorLambda(this.vpcSchemaInspectorLambda, this.config, schemaQuery, this.vpcLambdaRegion)
        : (await this.dbBuilder.raw(schemaQuery))[0];
    return this.queryToCSV(result);
  }

  public cleanup(): void {
    this.dbBuilder && this.dbBuilder.destroy();
  }
}

export function getMySQLSchemaQuery(databaseName: string): string {
  return `
SELECT DISTINCT
  INFORMATION_SCHEMA.COLUMNS.TABLE_NAME,
  INFORMATION_SCHEMA.COLUMNS.COLUMN_NAME,
  INFORMATION_SCHEMA.COLUMNS.COLUMN_DEFAULT,
  INFORMATION_SCHEMA.COLUMNS.ORDINAL_POSITION,
  INFORMATION_SCHEMA.COLUMNS.DATA_TYPE,
  INFORMATION_SCHEMA.COLUMNS.COLUMN_TYPE,
  INFORMATION_SCHEMA.COLUMNS.IS_NULLABLE,
  INFORMATION_SCHEMA.COLUMNS.CHARACTER_MAXIMUM_LENGTH,
  INFORMATION_SCHEMA.STATISTICS.INDEX_NAME,
  INFORMATION_SCHEMA.STATISTICS.NON_UNIQUE,
  INFORMATION_SCHEMA.STATISTICS.SEQ_IN_INDEX,
  INFORMATION_SCHEMA.STATISTICS.NULLABLE
FROM INFORMATION_SCHEMA.COLUMNS
LEFT JOIN INFORMATION_SCHEMA.STATISTICS ON INFORMATION_SCHEMA.COLUMNS.TABLE_NAME=INFORMATION_SCHEMA.STATISTICS.TABLE_NAME AND INFORMATION_SCHEMA.COLUMNS.COLUMN_NAME=INFORMATION_SCHEMA.STATISTICS.COLUMN_NAME
WHERE INFORMATION_SCHEMA.COLUMNS.TABLE_SCHEMA = '${databaseName}'
`;
}
