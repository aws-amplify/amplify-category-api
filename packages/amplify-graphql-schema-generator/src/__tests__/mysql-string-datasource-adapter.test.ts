import { StringDataSourceAdapter, MySQLStringDataSourceAdapter } from '../datasource-adapter';
import { Engine, Field, FieldType, Index, Model, Schema } from '../schema-representation';
import { generateGraphQLSchema, isComputeExpression } from '../schema-generator';
import { schemas } from './__utils__/schemas';
import { gql } from 'graphql-transformer-core';

class TestStringDataSourceAdapter extends StringDataSourceAdapter {
  public getTablesList(): string[] {
    return ['Test'];
  }

  public getFields(tableName: string): Field[] {
    return [];
  }

  public getPrimaryKey(tableName: string): Index | null {
    return null;
  }

  public getIndexes(tableName: string): Index[] {
    return [];
  }

  protected mapDataType(type: string, nullable: boolean): FieldType {
    return {
      kind: 'Scalar',
      name: 'String',
    };
  }

  protected parseSchema(schema: string): any[] {
    return [];
  }

  protected validateSchema(schema: any[]): void {}

  protected setSchema(schema: any[]): void {}

  protected setFields(fields: any[]): void {}

  protected setIndexes(indexes: any[]): void {}

  protected setTables(tables: any[]): void {}
}

describe('testStringDataSourceAdapter', () => {
  it('getModels call the default implementation', async () => {
    const adapter: StringDataSourceAdapter = new TestStringDataSourceAdapter('');
    adapter.getTablesList = jest.fn(() => ['Test']);
    adapter.getFields = jest.fn(() => []);
    adapter.getPrimaryKey = jest.fn();
    adapter.getIndexes = jest.fn(() => []);
    adapter.getModels();
    expect(adapter.getTablesList).toBeCalledTimes(1);
    expect(adapter.getFields).toBeCalledTimes(1);
    expect(adapter.getIndexes).toBeCalledTimes(1);
    expect(adapter.getPrimaryKey).toBeCalledTimes(1);
  });

  it('test mysql datatype mapping', () => {
    const schema = '';
    const adapter = new MySQLStringDataSourceAdapter(schema);
    expect(adapter.mapDataType('varchar', true, 'table', 'field', 'varchar(50)')).toEqual({
      kind: 'Scalar',
      name: 'String',
    });
    expect(adapter.mapDataType('char', true, 'table', 'field', 'varchar(50)')).toEqual({
      kind: 'Scalar',
      name: 'String',
    });
    expect(adapter.mapDataType('enum', true, 'table', 'field', 'enum("OPEN","CLOSED")')).toEqual({
      kind: 'Enum',
      name: 'table_field',
      values: ['OPEN', 'CLOSED'],
    });
    expect(adapter.mapDataType('bool', true, 'table', 'field', 'varchar(50)')).toEqual({
      kind: 'Scalar',
      name: 'Boolean',
    });
    expect(adapter.mapDataType('decimal', true, 'table', 'field', 'varchar(50)')).toEqual({
      kind: 'Scalar',
      name: 'Float',
    });
    expect(adapter.mapDataType('year', false, 'table', 'field', 'varchar(50)')).toEqual({
      kind: 'NonNull',
      type: {
        kind: 'Scalar',
        name: 'Int',
      },
    });
  });

  it('sets the correct models from a todo schema', () => {
    const adapter = new MySQLStringDataSourceAdapter(schemas.mysql.todo);
    expect(adapter.getModels()).toMatchSnapshot();
  });

  it('sets the correct models from a news schema', () => {
    const adapter = new MySQLStringDataSourceAdapter(schemas.mysql.news);
    expect(adapter.getModels()).toMatchSnapshot();
  });
});
