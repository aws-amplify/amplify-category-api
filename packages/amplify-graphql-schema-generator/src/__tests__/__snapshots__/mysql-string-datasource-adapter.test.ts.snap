// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testStringDataSourceAdapter sets the correct models from a news schema 1`] = `
Array [
  Model {
    "fields": Array [
      Object {
        "default": Object {
          "kind": "DB_GENERATED",
          "value": "NULL",
        },
        "length": "NULL",
        "name": "pub_id",
        "type": Object {
          "kind": "NonNull",
          "type": Object {
            "kind": "Scalar",
            "name": "Int",
          },
        },
      },
      Object {
        "default": Object {
          "kind": "DB_GENERATED",
          "value": "A",
        },
        "length": "1",
        "name": "pub_type",
        "type": Object {
          "kind": "NonNull",
          "type": Object {
            "kind": "Scalar",
            "name": "String",
          },
        },
      },
      Object {
        "default": Object {
          "kind": "DB_GENERATED",
          "value": "NULL",
        },
        "length": "1",
        "name": "placeholder",
        "type": Object {
          "kind": "NonNull",
          "type": Object {
            "kind": "Scalar",
            "name": "String",
          },
        },
      },
    ],
    "indexes": Array [
      Index {
        "fields": Array [
          "placeholder",
        ],
        "name": "NULL",
      },
    ],
    "name": "articles",
    "primaryKey": Index {
      "fields": Array [
        "pub_id",
        "pub_type",
      ],
      "name": "PRIMARY_KEY",
    },
  },
  Model {
    "fields": Array [
      Object {
        "default": Object {
          "kind": "DB_GENERATED",
          "value": "NULL",
        },
        "length": "NULL",
        "name": "pub_id",
        "type": Object {
          "kind": "NonNull",
          "type": Object {
            "kind": "Scalar",
            "name": "Int",
          },
        },
      },
      Object {
        "default": Object {
          "kind": "DB_GENERATED",
          "value": "CURRENT_TIMESTAMP",
        },
        "length": "NULL",
        "name": "comment_timestamp",
        "type": Object {
          "kind": "NonNull",
          "type": Object {
            "kind": "Scalar",
            "name": "AWSDateTime",
          },
        },
      },
      Object {
        "default": Object {
          "kind": "DB_GENERATED",
          "value": "NULL",
        },
        "length": "10",
        "name": "commenter_email",
        "type": Object {
          "kind": "NonNull",
          "type": Object {
            "kind": "Scalar",
            "name": "String",
          },
        },
      },
      Object {
        "default": Object {
          "kind": "DB_GENERATED",
          "value": "NULL",
        },
        "length": "30",
        "name": "comment_text",
        "type": Object {
          "kind": "NonNull",
          "type": Object {
            "kind": "Scalar",
            "name": "String",
          },
        },
      },
    ],
    "indexes": Array [
      Index {
        "fields": Array [
          "comment_text",
        ],
        "name": "NULL",
      },
    ],
    "name": "comments",
    "primaryKey": Index {
      "fields": Array [
        "pub_id",
        "comment_timestamp",
        "commenter_email",
      ],
      "name": "PRIMARY_KEY",
    },
  },
  Model {
    "fields": Array [
      Object {
        "default": Object {
          "kind": "DB_GENERATED",
          "value": "NULL",
        },
        "length": "NULL",
        "name": "pub_id",
        "type": Object {
          "kind": "NonNull",
          "type": Object {
            "kind": "Scalar",
            "name": "Int",
          },
        },
      },
      Object {
        "default": Object {
          "kind": "DB_GENERATED",
          "value": "NULL",
        },
        "length": "1",
        "name": "pub_type",
        "type": Object {
          "kind": "Scalar",
          "name": "String",
        },
      },
      Object {
        "default": Object {
          "kind": "DB_GENERATED",
          "value": "NULL",
        },
        "length": "64",
        "name": "pub_url",
        "type": Object {
          "kind": "NonNull",
          "type": Object {
            "kind": "Scalar",
            "name": "String",
          },
        },
      },
    ],
    "indexes": Array [
      Index {
        "fields": Array [
          "pub_id",
          "pub_type",
        ],
        "name": "publications_superkey",
      },
      Index {
        "fields": Array [
          "pub_url",
        ],
        "name": "pub_url",
      },
    ],
    "name": "publications",
    "primaryKey": Index {
      "fields": Array [
        "pub_id",
      ],
      "name": "PRIMARY_KEY",
    },
  },
  Model {
    "fields": Array [
      Object {
        "default": Object {
          "kind": "DB_GENERATED",
          "value": "NULL",
        },
        "length": "NULL",
        "name": "pub_id",
        "type": Object {
          "kind": "NonNull",
          "type": Object {
            "kind": "Scalar",
            "name": "Int",
          },
        },
      },
      Object {
        "default": Object {
          "kind": "DB_GENERATED",
          "value": "S",
        },
        "length": "1",
        "name": "pub_type",
        "type": Object {
          "kind": "NonNull",
          "type": Object {
            "kind": "Scalar",
            "name": "String",
          },
        },
      },
      Object {
        "default": Object {
          "kind": "DB_GENERATED",
          "value": "NULL",
        },
        "length": "1",
        "name": "placeholder",
        "type": Object {
          "kind": "NonNull",
          "type": Object {
            "kind": "Scalar",
            "name": "String",
          },
        },
      },
    ],
    "indexes": Array [
      Index {
        "fields": Array [
          "placeholder",
        ],
        "name": "NULL",
      },
    ],
    "name": "stories",
    "primaryKey": Index {
      "fields": Array [
        "pub_id",
        "pub_type",
      ],
      "name": "PRIMARY_KEY",
    },
  },
]
`;

exports[`testStringDataSourceAdapter sets the correct models from a todo schema 1`] = `
Array [
  Model {
    "fields": Array [
      Object {
        "default": Object {
          "kind": "DB_GENERATED",
          "value": "NULL",
        },
        "length": "NULL",
        "name": "id",
        "type": Object {
          "kind": "NonNull",
          "type": Object {
            "kind": "Scalar",
            "name": "Int",
          },
        },
      },
      Object {
        "default": Object {
          "kind": "DB_GENERATED",
          "value": "NULL",
        },
        "length": "255",
        "name": "title",
        "type": Object {
          "kind": "NonNull",
          "type": Object {
            "kind": "Scalar",
            "name": "String",
          },
        },
      },
      Object {
        "default": Object {
          "kind": "DB_GENERATED",
          "value": "NULL",
        },
        "length": "NULL",
        "name": "start_date",
        "type": Object {
          "kind": "Scalar",
          "name": "AWSDate",
        },
      },
      Object {
        "default": Object {
          "kind": "DB_GENERATED",
          "value": "NULL",
        },
        "length": "NULL",
        "name": "due_date",
        "type": Object {
          "kind": "Scalar",
          "name": "AWSDate",
        },
      },
    ],
    "indexes": Array [
      Index {
        "fields": Array [
          "due_date",
          "start_date",
          "title",
        ],
        "name": "NULL",
      },
    ],
    "name": "Todo",
    "primaryKey": Index {
      "fields": Array [
        "id",
      ],
      "name": "PRIMARY_KEY",
    },
  },
]
`;
